fix issue 193

HerrDingenz opened the pull request:
#### Describe the issue (if no issue has been made)
As stated in Issue #193 there is a need for the implementation of file loggers.

#### PR impact _(Make sure to add [closing keywords](https://help.github.com/en/articles/closing-issues-using-keywords))_
List of related issues/PRs this will solve:

 Impact                  | Issue/PR
------------------------ | ------
Issues this solves       | Closes #193
Other PRs this solves    | None or #number(s)

#### Proposed fix _(Make sure you've read [on how to contribute](https://github.com/TheCherno/Hazel/blob/master/.github/CONTRIBUTING.md) to Hazel)_
I implemented a file logger for both Hazel Core and Hazel Client.
The Hazel Core file logger will create a file named "hazel-log.txt".
The Client file logger will create a file named "client-log.txt".

The console loggers will only be available in Debug builds while the file loggers will be available in Debug and Release builds.

NOTE:
I changed an if-statement in Instrumentor.h --- It now checks if the file logger is initialized.

#### Additional context
I tested and ran both Debug and Release build at it seems to work like a charm

HerrDingenz commented:
Thanks for your review.

The idea using sinks is pretty nice but:

I didn't use sinks because i wanted to use different patterns for both loggers. I think all that timestamp information just inflates the file logs - and we really don't need that information in there.

For now the macros do the trick - even though we have 4 different loggers now (at least in debug builds). In release builds there are only 2 loggers (the file loggers obviously)

Regarding the filenames you are right - they should fit the other style of hazel filenames.

HerrDingenz commented:
Ok .. I changed it to use sinks.

But i am not really happy with the result, because now we have the timestamps AND the logger name inside the log files.

Not only increases this the file size - it is useless information!
We shouldn't care about what time the log message occured - cause the files get created per session anyway.
And since we have 2 different files (one for core and one for client) the logger name inside the files is useless in there, too.

For output to the console this information was useful (i mean the name of the logger) and we need that preset there because of color formatting. But for the file loggers it is just data garbage nobody needs inside those files.

And as i stated in my last comment already: we can't use different presets for the loggers anymore when switching to multiple sinks for one logger.

CleverSource commented on Hazel/src/Hazel/Core/Log.cpp:
```diff
@@ -1,21 +1,36 @@
 #include "hzpch.h"
-#include "Hazel/Core/Log.h"
+//#include "Hazel/Core/Log.h"		included in hzpch.h already
```
This should just be removed, someone can easily find out that it's included in the pch without this here.

LovelySanta replied:
Cherno mensioned this in one of his videos, we should not rely in Hazel files being included in the precompiled header. The reason we added it to the pch in the first place was because we won't update it often. And especialy in this case where the cpp file should (always) include its header file, so can you please put it back?

HerrDingenz replied:
that is done already .. the include parts of the cpp file are as they have been before my pull request

HerrDingenz replied:
i already put that back yesterday ;)

CleverSource commented on Hazel/src/Hazel/Core/Log.cpp:
```diff
@@ -1,21 +1,36 @@
+		s_ConsoleSink = std::make_shared<spdlog::sinks::stdout_color_sink_mt>();
+
+		std::vector<spdlog::sink_ptr> coreSinks;
+		coreSinks.emplace_back(s_ConsoleSink); // VS debug console
```
This isn't necessarily required, but it won't always be the Visual Studio debug console, which somewhat makes this comment useless.

HerrDingenz commented:
Thanks for your review again, LovelySanta...

I made some changes to the code.

Now we are using just one sink and only one log file (for both - client and engine log messages).
The log file will simply just be named "Hazel.log" (for now).

I changed the names of the loggers to "ENGINE" and "CLIENT" and added "%L" to the log patterns. I did that to to make it look more straight inside the log file...

ENGINE I: This is an info.
CLIENT C: This is critical!

is more readable than:

HAZEL: This is an info.
APP: This is critical!

or:

HAZEL info: This is an info.
APP critical: This is critical!

The output to the console stays the same as before.

HerrDingenz commented:
trying to optimize it as much as i can .. this is my last commit to this PR.

if you're not happy with it then just close it

CleverSource commented:
I think it should be formatted ``[Severity] Level: Text`` so it could be ``[Critical] Hazel: Some critical text here.``

HerrDingenz commented:
Done.

HerrDingenz commented:
I added "*.log" to the ".gitignore" file to make sure the log file doesn't get pushed to the repository.

LovelySanta commented on Hazel/src/Hazel/Core/Log.cpp:
```diff
@@ -10,12 +11,22 @@ namespace Hazel {
+		s_ClientLogger = std::make_shared<spdlog::logger>("APP", begin(logSinks), end(logSinks));
+		spdlog::register_logger(s_ClientLogger);
 		s_ClientLogger->set_level(spdlog::level::trace);
+
```
Remove this last empty line on the function

HerrDingenz replied:
sure

LovelySanta commented on .gitignore:
```diff
@@ -2,6 +2,7 @@
 **/bin/
 !vendor/bin/premake/
 bin-int/
+*.log
```
Maybe make a new group (aka a comment line above) for hazel files?
```.gitignore
# Hazel files
*.log
```

HerrDingenz replied:
done

HerrDingenz commented:
well then i guess we are ready to merge :wink:

TheCherno commented:
Thank you to everyone for their contributions to this, I'd love to merge this in, however, I have a few issues:
- Logged file doesn't seem to contain any timestamps, like the output log does. This would be useful to have
- Flushing the file seems to be tied to the engine shutting down cleanly, which isn't always the case - I would love to have a log file of what happened if the engine crashed, but currently I would just have a blank Hazel.log file

Could we get these issues resolved? I'm open to discussion.

HerrDingenz commented:
Adding the timestamps back in can be done by just adding them back to the pattern of the file sink. We stripped them out because the log file gets created per session anyway so the timestamps might not be really that necessary in the log file.

For the log file not being filled if the engine crashes that is really an issue. In my first version i did the flushing myself inside the macros everytime a message got put to the file.

HerrDingenz commented:
Timestamps are added back to the log file.
Flush level for the loggers is set to "trace" - so now spdlog will flush the contents at every message and we will have data inside the log file even if Hazel crashes.
